File: problem1B.java
class longestIncreasingSubsequence {
    public int lengthOfLIS(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        
        int n = nums.length;
        int[] dp = new int[n];
        dp[0] = nums[0];
        int len = 1;
        
        for(int i = 1; i < n; i++){
            if(nums[i] > dp[len - 1]){
                dp[len] = nums[i];
                len++;
            }
            else{
                int idx = binarySearch(dp, 0, len - 1, nums[i]);
                dp[idx] = nums[i];
            }
        }
        return len;
    }
    
    private int binarySearch(int[] dp, int low, int high, int target){
        while(low <= high){
            int mid = low + (high - low) / 2;
            if(dp[mid] == target){
                return mid;
            }
            else if(target < dp[mid]){
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return low;
    }
}

//time complexity O(n * log n) where n is length of given array
//space complexity O(n)

File: Problem1b.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:

        # The code implements the "Longest Increasing Subsequence" problem using a binary search-based approach.
        # Time complexity - O(n log n) due to the binary search operation
        # Space complexity - O(n) due to the arr array
        if nums is None or len(nums) == 0:
            return 0

        n = len(nums)
        # Initialize an array to store the longest increasing subsequence ending at each index
        arr = [0] * n
        arr[0] = nums[0]
        longest = 1  # Initialize the length of the longest increasing subsequence

        # Binary search function to find the insertion point of a value in the sorted array
        def binarySearch(arr, low, high, target):
            while low <= high:
                mid = low + (high - low) // 2
                if arr[mid] == target:
                    return mid
                elif arr[mid] > target:
                    high = mid - 1
                else:
                    low = mid + 1
            return low

        for i in range(1, n):
            if nums[i] <= arr[longest - 1]:
                # Find the insertion point
                bsIdx = binarySearch(arr, 0, longest - 1, nums[i])
                # Insert the value at the appropriate position
                arr[bsIdx] = nums[i]
            else:
                # Update the last element of the longest subsequence
                arr[longest] = nums[i]
                longest += 1

        return longest
